src/Controller/ControllerProject.php

```diff
@@ -273,4 +284,77 @@ public function __construct($entityManager, $user)
             }
         );
     }
+
+    public function assignRelatedExercicesAndTestsToStudent($project,$projectDuplicated){
+       
+        $this->entityManager->getConnection()->beginTransaction();
+        $success = true;
+        try{
+            // get python exercice
+            $pythonExerciseFound = $this->entityManager
+                ->getRepository(ExercisePython::class)
+                ->findOneByProject($project);
+
+            if($pythonExerciseFound){
+
+                // we create and save the exercise with the related project
+                $duplicatedPythonExercise = new ExercisePython($pythonExerciseFound->getFunctionName());
+                $duplicatedPythonExercise->setProject($projectDuplicated);
+                $this->entityManager->persist($duplicatedPythonExercise);          
+
+                // get python test related to this exercise in python_tests table
+                $pythonTest = $this->entityManager
+                    ->getRepository(UnitTests::class)
+                    ->findOneByExercise($pythonExerciseFound);
+
+                if(!$pythonTest) $success = false;
+            
+                // we create and save the python test with the related exercise
+                $duplicatedPythonTest = new UnitTests();
+                $duplicatedPythonTest->setExercise($duplicatedPythonExercise);
+                $duplicatedPythonTest->setHint($pythonTest->getHint());
+                $this->entityManager->persist($duplicatedPythonTest);
+
+                // get unit tests inputs related to this unit test in python_tests_inputs
+                $pythonTestInputs = $this->entityManager
+                    ->getRepository(UnitTestsInputs::class)
+                    ->findByUnitTest($pythonTest);
+
+                if(!$pythonTestInputs) $success = false;
+
+                foreach($pythonTestInputs as $pythonTestInput){
+                    $duplicatedTestInput = new UnitTestsInputs();
+                    $duplicatedTestInput->setUnitTest($duplicatedPythonTest);
+                    $duplicatedTestInput->setValue($pythonTestInput->getValue());
+                    $this->entityManager->persist($duplicatedTestInput);
+                }
+                
+                $pythonTestOutputs = $this->entityManager
+                    ->getRepository(UnitTestsOutputs::class)
+                    ->findByUnitTest($pythonTest);
+                
+                if(!$pythonTestOutputs) $success = false;
```

Comment:
Between lines 310, 323 and 336. You are checking whether success or not, and every time success has value you don't check against it. Here you are not respecting two principles of software engineering: 

1. Fail Fast - there is no meaning of proceeding with the execution if one of the steps failed. Loss of resources if you fail in one step and keep going (entityManager is connected to the database, you are taking resources from the database that you should not)
2. DRY Don't repeat yourself 

```             
            try {
                // we create and save the exercise with the related project
                $duplicatedPythonExercise = new ExercisePython($pythonExerciseFound->getFunctionName());
                $duplicatedPythonExercise->setProject($projectDuplicated);
                $this->entityManager->persist($duplicatedPythonExercise);          

                // get python test related to this exercise in python_tests table
                $pythonTest = $this->entityManager
                    ->getRepository(UnitTests::class)
                    ->findOneByExercise($pythonExerciseFound);
                 
                if (!$pythonTest) throw new Exception("something wrong with the test") //FAIL FAST
                ...
                if(!$pythonTestOutputs) throw new Exception("something wrong with the test output") //FAIL FAST
                ...
                return true;
             }   
            catch(\Exception $e){
               $this->entityManager->getConnection()->rollback();
               return false;
            } 
                  
```